### DISCUSSION TITLE
charmm force field

### DISCUSSION THREAD

**Discussion by Sizhang92190** [OpenFreeEnergy/openfe#659] [Q&A]
I've been trying to use `openfe` recently to perform RBFE calculations on a receptor (containing an ADP residue) + a couple of ligands. Due to the presence of ADP, I've had to replace the amber force field with the charmm force field in the script, `gufe/settings/models.py`.
```
i.e., 
forcefields: list[str] = [
  "charmm/charmm36_nowaters.xml",
  "charmm/waters_ions_default.xml",
  ] 
```
However, the calculation failed in the parameterization step due to a memory issue. (It still failed even if the memory was increased.) @mikemhenry Any ideas? 
Would appreciate any suggestions! If necessary, I can upload a few files later.

**Comment by IAlibay** [comment]
Thanks so much for reporting this @Sizhang92190 !

Admittedly, I don't think we've tried using the charmm FFs yet, mostly because of small molecule FF compatibility issues and also I'm not 100% sure that short range shifted cutoffs are working as expected when building the hybrid topology. That being said we should definitely try to support it if we can.

Do you happen to still have the error trace for the memory failure? That would help up try to work out where things are going wrong.

**Reply by Sizhang92190** [reply]
Thanks for your reply! I guess you're right. There should be FF compatibility issues between charmm FFs for proteins and openff or gaff for small molecules. It then causes the simulation to get stuck in the `'Adding Forces'` step when building the `hybrid_factory`.

**Comment by mikemhenry** [comment]
Something you can try to debug this issue is to use https://github.com/openmm/openmmforcefields directly to parametrize a system with charmm without having to run things through the openFE pipeline

**Reply by Sizhang92190** [reply]
Thanks, Mike! I tried to build a `hybrid_factory` after I separately parameterized systemA and B (receptor + ligand) with charmm ff. It failed with the error below:
```
/usr/mambaforge/envs/openfe_env/lib/python3.11/site-packages/openfe/protocols/openmm_rfe/_rfe_utils/relative.py", line 358, in _check_unknown_forces
    raise ValueError(errmsg)
ValueError: Unknown forces {'CustomNonbondedForce', 'CMAPTorsionForce', 'CustomTorsionForce'} encountered in old system
```
I could get a hybrid_factory if I naively added these 3 forces to `known_forces`. But the system clashed in either the minimization step or the equilibration step.

Not sure how to properly deal with this issue. Do you have any suggestions?

**Reply by mikemhenry** [reply]
@IAlibay will have to comment on this, looking at the code https://github.com/OpenFreeEnergy/openfe/blob/5ddaad210ae9b41e5272ed9cd4e619f14541dac9/openfe/protocols/openmm_rfe/_rfe_utils/relative.py#L326-L330

it looks like we have a check to make sure that we only copy forces that we expect to handle, are you able to run a "normal" MD run with the system that you parameterized with charmm? Before we try to hard to make it work/support this, I just want to check these system can run MD before we worry about any alchemy.

**Reply by Sizhang92190** [reply]
Hi Mike, yes, I can successfully run a "normal" MD for system A or B using gromacs or openmm (seems openmm support these forces). It looks like the funcs of those 3 froces also need to be added to `relative.py.`

**Reply by mikemhenry** [reply]
I am not sure how to interpolate the forces that get added when charmm is used, so you would need to modify `relative.py`. Given you can can run MD then it sounds like these are stable... If you manually add the forces to the `known_forces` you said it ends up crashing, is it crash because of a NaN or do you hit another error?

**Reply by Sizhang92190** [reply]
Yea, I'm trying to modify `relative.py`, but I have no idea how to properly create funcs for those forces right now.

If I just add them to the `known_forces`, I can get the NaN error below: (tried several times and got the same thing)

```
Running minimization
Running equilibration phase
[/usr/mambaforge/envs/openfe_env/lib/python3.11/site-packages/openmmtools/mcmc.py:723]: UserWarning: Particle coordinate is NaN.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan
  warnings.warn(str(e))
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 2 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 3 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 4 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 5 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 6 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 7 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 8 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 9 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 10 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 11 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 12 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 13 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 14 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 15 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 16 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 17 attempts of integration with move LangevinDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 18 attempts of integration with move LangevinDynamicsMove Attempting a restart...
ERROR:openmmtools.mcmc:Potential energy is NaN after 19 attempts of integration with move LangevinDynamicsMove Trying to reinitialize Context as a last-resort restart attempt...
ERROR:openmmtools.mcmc:Potential energy is NaN after 20 attempts of integration with move LangevinDynamicsMove
CRITICAL:openmmtools.multistate.multistatesampler:Propagating replica 0 at state 0 resulted in a NaN!
The state of the system and integrator before the error were saved in complex_jobs/nan-error-logs
```

**Reply by IAlibay** [reply]
@Sizhang92190 - unfortunately as @mikemhenry alludes to, the issue is down to the fact that the HybridTopologyFactory does not know how to properly account for these extra forces. So what you're getting here is a system generated that probably just doesn't include these additional forces at all (hence the NaNs).

Adding specific means of dealing with these forces is not a simple task - it's something that would take a decent chunk of development time (order of month+), particularly for the `CustomNonbondedForce` (which I assume to be some kind of truncated shifted nonbonded interaction and would require effort to work out how to properly deal with in an alchemical sense).

It's something we want to deal with one day, but we unfortunately can't dedicate time to it currently. We can probably advise on what could be done here if you're willing to implement a fix yourself, but my short to medium term suggestion would be to avoid non-AMBER force fields, if possible.

**Reply by IAlibay** [reply]
Our approach to cofactors (like ADP) has been to parameterize them via OpenFF as just an additional spectator small molecule. This seems to have worked reasonably well in most of our cases.

**Reply by Sizhang92190** [reply]
Thanks for the detailed explanation! @IAlibay `CMAP atomtypes` would also be an issue if I use `charmmff` with the `openfe` pipeline.

About using spectator, it's a good idea! But I see that it's now only available in the `perses` functions that requires an OpenEye license. I don't have one.

Finally, I figured out one way to run RBFE for a system containing ADP and phosphate with the current `openfe` pipeline. There're two files that need to be modified: 1.) Add funcs that generates bond orders and types of 'nonstandard' residues (e.g., ADP) to 
`gufe/components/proteincomponent.py`, 2.) Use `extended_omm_forcefield` instead of `omm_forcefield` in `openfe/protocols/openmm_rfe/equil_rfe_methods.py`. `extended_omm_forcefield` contains the generated SMIRNOFF parameters for ADP.

```
    extended_omm_forcefield=system_generator.forcefield 
    from openff.toolkit import Molecule
    molecules = Molecule.from_file(residue_template_fname)
    from openmmforcefields.generators import SMIRNOFFTemplateGenerator
    smirnoff = SMIRNOFFTemplateGenerator(molecules=molecules)
    smirnoff.add_molecules(molecules)
    extended_omm_forcefield.registerTemplateGenerator(smirnoff.generator)
```
Since I wasn't sure which `openfe` or/and `gufe` functions I needed to modify, I just simply added above code to the second file. (If you have any suggestions, that would be helpful!)

To let the above scripts work, I just need to provide `residue_template_fname` with a sdf file along with a template residue (e.g., ADP). 

Attached are those two modifed files and the template residue file. Just in case if someone has the similar goal. Btw, I use `openfe-0.12.0`.
[files.zip](https://github.com/OpenFreeEnergy/openfe/files/13638447/files.zip)


Please let me know if I make a mistake. Thanks!

**Reply by IAlibay** [reply]
@Sizhang92190 there's a better way to do this within `openfe `but it seems like we haven't advertised it very well. It might be a lot easier to discuss in a call if that's something you want to have?

You may need to use a slightly newer version of `openfe`.

The short story is that you only need to pass through extra `SmallMoleculeComponent`s to your protocol and these will be handled as cofactors.

**Reply by Sizhang92190** [reply]
An extra `SmallMoleculeComponent` sounds like a better and simpler way. I really should have known before I tried those things. 
I'd like to discuss it with you if you email me an invitation link. (My email: sizhang@temple.edu)

**Reply by IAlibay** [reply]
Ok, I've just sent you an email with some proposed meeting times.
